libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/map/MapEventDelegate.kt

```diff
@@ -0,0 +1,97 @@
+import com.mapbox.navigation.utils.internal.InternalJobControlFactory
+import com.mapbox.navigation.utils.internal.ifNonNull
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.launch
+
+class MapEventDelegate(private val mapView: MapView) {
+
+    private val jobControl = InternalJobControlFactory.createMainScopeJobControl()
+
+    private val _mapStyleSink: MutableSharedFlow<Style> = MutableSharedFlow(
+        replay = 1,
+        onBufferOverflow = BufferOverflow.DROP_OLDEST,
+        extraBufferCapacity = 1
+    )
+    val mapStyleUpdates: Flow<Style> = _mapStyleSink
+
+    private val _mapLongClickSink: MutableSharedFlow<Point> = MutableSharedFlow(
+        extraBufferCapacity = 1,
+        onBufferOverflow = BufferOverflow.DROP_OLDEST
+    )
+    val mapLongClicks: Flow<Point> = _mapLongClickSink
+
+    private val _mapClickSink: MutableSharedFlow<Point> = MutableSharedFlow(
+        extraBufferCapacity = 1,
+        onBufferOverflow = BufferOverflow.DROP_OLDEST
+    )
+    val mapClicks: Flow<Point> = _mapClickSink
+
+    private val _positionChangeSink: MutableSharedFlow<Point> = MutableSharedFlow()
+    val positionChanges: Flow<Point> = _positionChangeSink
+
+    private val _mapMovementsSink: MutableSharedFlow<MotionEvent> = MutableSharedFlow(
+        extraBufferCapacity = 1,
+        onBufferOverflow = BufferOverflow.DROP_OLDEST,
+    )
+    val mapMovements: Flow<MotionEvent> = _mapMovementsSink
+
+    private val onStyleLoadedListener = OnStyleLoadedListener {
+        ifNonNull(mapView.getMapboxMap().getStyle()) {
+            _mapStyleSink.tryEmit(it)
+        }
+    }
+
+    private val onMapLongClickListener = OnMapLongClickListener { point ->
+        _mapLongClickSink.tryEmit(point)
+        false
+    }
+
+    private val onMapClickListener = OnMapClickListener { point ->
+        _mapClickSink.tryEmit(point)
+        false
+    }
+
+    private val onPositionChangedListener = OnIndicatorPositionChangedListener { point ->
+        jobControl.scope.launch {
+            _positionChangeSink.emit(point)
+        }
+    }
+
+    private val onMoveListener = object : OnMoveListener {
+        override fun onMove(detector: MoveGestureDetector): Boolean {
+            return false
+        }
+
+        override fun onMoveBegin(detector: MoveGestureDetector) { }
+
+        override fun onMoveEnd(detector: MoveGestureDetector) {
+            _mapMovementsSink.tryEmit(detector.currentEvent)
+        }
+    }
+
+    init {
+        mapView.getMapboxMap().addOnStyleLoadedListener(onStyleLoadedListener)
+        mapView.getMapboxMap().addOnMapLongClickListener(onMapLongClickListener)
+        mapView.getMapboxMap().addOnMapClickListener(onMapClickListener)
+        mapView.location.addOnIndicatorPositionChangedListener(onPositionChangedListener)
+        mapView.getMapboxMap().addOnMoveListener(onMoveListener)
+    }
```

Comment:
Discussed offline but making note. We're making everything lifecycle aware, so they can "remove" their listeners. I'll let you decide what to do today, because this object has the same lifecycle as the MapView and does not appear to leak.

But we have a ticket to make the MapView "injectable" https://github.com/mapbox/navigation-sdks/issues/1473. So we should be able to add and remove this delegate. It would be nice to be able to update the map with a new map. We also may need to pass around styles.

Proposing, making this class extend `MapboxNavigationObserver` (for lifecycle reasons). Another option is `DefaultLifecycleObserver`, but `removeObserver` does not call destruction events so that would probably be a temporary solution too.

Reply:
not blocking. but it's an issue that will need to be addressed if we want to making MapView swappable